### ISSUE TITLE
recursively select files

### ISSUE THREAD

**Issue by mohkale** [gokcehan/lf#371]
Today I had a simple problem. I wanted to select all files at depth 2.

eg.

```
foo/
foo/bar
foo/baz
foo/bag
...
bar/
bar/foo
bar/baz
bar/bag
...
```

but lf doesn't seem to support recursive globs I.E. `:glob-select **/*`. In ranger I'd normally do `:flat 0` and then filter out directories so I can select the files, but `lf` doesn't have that feature yet so is there another recommended way?

I tried the subprocess approach:

```
find -type f -maxdepth 2 -mindepth 2 -exec lf -remote "send $id :select \"{}\"" ';'
```

but that spawns a tonne of lf processes and seems to hang indefinitely if you try to do too many files at once.

**Comment by gotroyb127** [comment]
Here's what i use:

````shell
cmd SelRecurs %{{
	IFS=$'\n'
	[[ ! -d $f ]] && exit 0
	for d in $(find $f -type d); do
		lf -remote "send $id select '$d'"$'\n'\
\	\	\	"send $id open"$'\n'\
\	\	\	"send $id invert"
	done
	lf -remote "send $id select '$f'"
}}
````

This selectes every subdirectory of the selected directory, enters it selects (actually toggles selection on) everything inside that (and then repeat to a next one, if there is any).
(The `IFS=$'\n'` is used because I use the default `ifs` value `"\n"` )

It have worked for me very well for what I use it and it seems to be effective even for directories with too many subdirectories.

You can modify it to fit your needs.

**Comment by gokcehan** [comment]
@mohkale Easiest way to fix this would be to add recursive capabilities to our current `glob-select` command as you mention much like globstar in bash.

**Comment by mohkale** [comment]
It'd be nice if we could simply use `:glob-select */*`. That way, at least for simple cases, we could specify how many directories deep it should be able to go. 

- `:glob-select */*` selects all directories at depth 2
- `:glob-select */*/*` selects all directories at depth 3
- `:glob-select */*/*/*` selects all directories at depth 4
...

**Comment by gokcehan** [comment]
@mohkale I'm hesitant to add a custom syntax for this purpose. Is it used somewhere?

**Comment by mohkale** [comment]
I've never seen something like this explicitly demonstrated, but I believe its the expected behavior with globs. It seems to work in: 

- ruby `Dir.glob('*/*')`
- python `from pathlib import Path as p; list(p('.').glob('*/*'))`
- node `var glob = require("glob"); g.glob('*/*', {}, function(err, files) { console.log(files) })`

I presume if you allow recursive globs than globs like this will also be enabled, but I'm not at all familiar with golang so I can't say for sure.

**Comment by gokcehan** [comment]
@mohkale So then it is used somewhere. I also realized this already works in Go as well. But we're using the base name of a file in `glob-select` implementation and we only check the files in the current directory. We need to walk the directory to implement this though we should somehow do this efficiently in a smart way. It may be a little tricky to implement.

**Comment by jpggithub** [comment]
The `flat` command in `ranger` is very useful, not only for selecting files but also to open them or just to have a glimpse of them. If you have a `flat` command, you can do many things. 
If you only modify  \`glob-select\`, you can only... select files ;-)

**Comment by joelim-work** [comment]
So apparently this works:

```sh
cmd select-glob &{{
    lf -remote "send $id :unselect; toggle $(echo $1)"
}}
```

Use `:select-glob */*` to select all files at depth 2

- The `$1` should not be quoted, otherwise the glob expression will be treated literally instead of being expanded
- `echo $1` just prints the filenames of the expanded glob expression separated by spaces
- The `toggle` command can accept multiple filenames, hence the use of the `echo` command above

This is for the default shell `sh`. For other shells like `bash`/`zsh`, some additional settings may be required. Also this example is simple and assumes that filenames don't contain spaces, otherwise they would have to be escaped when passing them to the `toggle` command.

**Comment by joelim-work** [comment]
So I finally got around to adding this as a [tip](https://github.com/gokcehan/lf/wiki/Tips#select-files-matching-recursive-glob) in the wiki.

In my opinion, this is how glob selection should have been implemented (as external configuration, in line with `lf`'s minimalistic design) instead of a builtin command. But since `glob-select` already exists, I think it's better to just keep it, since removing it will constitute a breaking change.

**Comment by jpggithub** [comment]
Yes it answers to the initial question but IMHO it is far less useful that the `flat` function of `ranger`.

**Comment by joelim-work** [comment]
There's already an open issue for `flat`, see #227

**Comment by jpggithub** [comment]
Thanks @joelim-work
